lib/dal/src/action.rs

```diff
@@ -759,6 +776,142 @@ impl Action {
+    ) -> ActionResult<()> {
+        let head_change_set_id = ctx.get_workspace_default_change_set_id().await?;
+        // if we're already on head, just enqueue it
+        if ctx.change_set_id() == head_change_set_id {
+            Self::enqueue_refresh(ctx, component_id, false).await
+        } else {
+            // Not on head, so build the head dal ctx
+            let head_ctx = ctx.clone_with_head().await?;
+            // if the component exists on head, enqueue the refresh action there
+            if Component::exists_by_id(&head_ctx, component_id).await? {
+                Self::enqueue_refresh(&head_ctx, component_id, false).await
+            }
+            // otherwise, if the component doesn't have a resource, just enqueue it
+            // (this is a backend guard, the button will be hidden from the user)
+            else if Component::resource_by_id(ctx, component_id)
+                .await?
+                .is_none()
+            {
+                Self::enqueue_refresh(ctx, component_id, false).await
+            }
+            // last case, component doesn't exist on head, but has a resource (which can only be true if we ran an import or other mgmt func)
+            // so we can enqueue and dispatch right away
+            else {
+                Self::enqueue_refresh(ctx, component_id, true).await
+            }
+        }
+    }
+
+    async fn enqueue_refresh(
+        ctx: &DalContext,
+        component_id: ComponentId,
+        should_dispatch: bool,
+    ) -> ActionResult<()> {
+        let schema_variant_id = Component::schema_variant_id(ctx, component_id).await?;
+        let refresh_actions = SchemaVariant::find_action_prototypes_by_kind(
+            ctx,
+            schema_variant_id,
+            ActionKind::Refresh,
+        )
+        .await?;
+        if refresh_actions.len() > 1 {
+            return Err(ActionError::UnexpectedNumberOfActionKinds(
+                ActionKind::Refresh,
+                schema_variant_id,
+            ));
+        }
+        if let Some(&prototype_id) = refresh_actions.first() {
+            let maybe_duplicate_action =
+                Action::find_for_kind_and_component_id(ctx, component_id, ActionKind::Refresh)
+                    .await?;
+
+            // if we're not dispatching and we've got a refresh action enqueued already
+            // we'll make this a 'retry' as this whole function exists to interpret what the user wanted
+            // which just means if the existing refresh action is Failed/On Hold, queue it up!
+            if !maybe_duplicate_action.is_empty() && !should_dispatch {
+                if maybe_duplicate_action.len() > 1 {
+                    return Err(ActionError::UnexpectedNumberOfActionsEnqueuedForComponent(
+                        ActionKind::Refresh,
+                        component_id,
+                    ));
+                }
+                if let Some(action_id) = maybe_duplicate_action.first() {
```

Comment:
[`Itertools::at_most_one`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.at_most_one) could probably tidy this up a bit.